common/cache/lru.go

```diff
@@ -291,23 +291,25 @@ func (c *lru) putInternal(key interface{}, value interface{}, allowUpdate bool)
 	}
 
 	elt := c.byKey[key]
+	// If the entry exists, check if it has expired or update the value
 	if elt != nil {
-		entry := elt.Value.(*entryImpl)
-		if c.isEntryExpired(entry, time.Now().UTC()) {
+		existingEntry := elt.Value.(*entryImpl)
+		if c.isEntryExpired(existingEntry, time.Now().UTC()) {
 			// Entry has expired
 			c.deleteInternal(elt)
 		} else {
-			existing := entry.value
+			existing := existingEntry.value
 			if allowUpdate {
-				entry.value = value
+				c.currSize -= existingEntry.Size()
+				existingEntry.value = value
 				if c.ttl != 0 {
-					entry.createTime = time.Now().UTC()
+					existingEntry.createTime = time.Now().UTC()
 				}
 			}
```

Comment:
If allowUpdate is false, then you need to undo the update to currSize made on line 285 (in an else branch here).

I think it would be much easier to see that this is correct if you made the update to currSize at the same time as adding/removing/modifying an entry.